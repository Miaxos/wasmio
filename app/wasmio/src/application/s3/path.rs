//! Part of it was taken from `s3-server` crate.
use std::net::IpAddr;

use axum::async_trait;
use axum::extract::{FromRequestParts, Path};
use axum::http::request::Parts;
use ulid::Ulid;

use super::errors::{S3Error, S3ErrorCodeKind, S3HTTPError};

#[derive(Debug)]
pub enum S3Path {
    Root,
    Bucket { bucket: String },
    Object { bucket: String, key: String },
}

impl S3Path {
    pub fn from_part(
        request_id: &Ulid,
        parts: &Parts,
    ) -> Result<Self, S3HTTPError> {
        let path = parts.uri.path();
        let path = urlencoding::decode(path).map_err(|e| {
            S3HTTPError::custom(
                "",
                request_id.to_string(),
                S3ErrorCodeKind::InvalidURI,
            )
        })?;
        let path = Self::try_from_path(&path).map_err(|e| {
            S3HTTPError::custom(&path, request_id.to_string(), e)
        })?;

        Ok(path)
    }

    /// See [bucket nameing rules](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html#bucketnamingrules)
    #[must_use]
    pub fn check_bucket_name(name: &str) -> bool {
        if !(3_usize..64).contains(&name.len()) {
            return false;
        }

        if !name.as_bytes().iter().all(|&b| {
            b.is_ascii_lowercase()
                || b.is_ascii_digit()
                || b == b'.'
                || b == b'-'
        }) {
            return false;
        }

        if name
            .as_bytes()
            .first()
            .map(|&b| b.is_ascii_lowercase() || b.is_ascii_digit())
            != Some(true)
        {
            return false;
        }

        if name
            .as_bytes()
            .last()
            .map(|&b| b.is_ascii_lowercase() || b.is_ascii_digit())
            != Some(true)
        {
            return false;
        }

        if name.parse::<IpAddr>().is_ok() {
            return false;
        }

        if name.starts_with("xn--") {
            return false;
        }

        true
    }

    /// The name for a key is a sequence of Unicode characters whose UTF-8
    /// encoding is at most 1,024 bytes long. See [object keys](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-keys)
    #[must_use]
    pub const fn check_key(key: &str) -> bool {
        key.len() <= 1024
    }

    /// Parse a path-style request
    /// # Errors
    /// Returns an `Err` if the s3 path is invalid
    pub fn try_from_path(path: &str) -> Result<Self, S3Error> {
        let path = if let Some(("", x)) = path.split_once('/') {
            x
        } else {
            return Err(S3ErrorCodeKind::InvalidURI.into());
        };

        if path.is_empty() {
            return Ok(S3Path::Root);
        }

        let (bucket, key) = match path.split_once('/') {
            None => (path, None),
            Some((x, "")) => (x, None),
            Some((bucket, key)) => (bucket, Some(key)),
        };

        if !Self::check_bucket_name(bucket) {
            return Err(S3ErrorCodeKind::InvalidBucketName.into());
        }

        let key = match key {
            None => {
                return Ok(S3Path::Bucket {
                    bucket: bucket.to_string(),
                })
            }
            Some(k) => k,
        };

        if !Self::check_key(key) {
            return Err(S3ErrorCodeKind::KeyTooLongError.into());
        }

        Ok(Self::Object {
            bucket: bucket.to_string(),
            key: key.to_string(),
        })
    }

    #[must_use]
    pub const fn is_root(&self) -> bool {
        matches!(*self, Self::Root)
    }

    #[must_use]
    pub const fn is_bucket(&self) -> bool {
        matches!(*self, Self::Bucket { .. })
    }

    #[must_use]
    pub const fn is_object(&self) -> bool {
        matches!(*self, Self::Object { .. })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_s3_path_root() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("/"));
    }

    #[test]
    fn parse_s3_path_bucket() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("/bucket"));
    }

    #[test]
    fn parse_s3_path_bucket_2() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("/bucket/"));
    }

    #[test]
    fn parse_s3_path_object() {
        insta::assert_debug_snapshot!(S3Path::try_from_path(
            "/bucket/dir/object"
        ));
    }

    #[test]
    fn parse_s3_path_fail() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("asd"));
    }

    #[test]
    fn parse_s3_path_fail_2() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("a/"));
    }

    #[test]
    fn parse_s3_path_fail_3() {
        insta::assert_debug_snapshot!(S3Path::try_from_path("/*"));
    }

    #[test]
    fn parse_s3_path_fail_4() {
        let too_long_path = format!("/{}/{}", "asd", "b".repeat(2048).as_str());
        insta::assert_debug_snapshot!(S3Path::try_from_path(&too_long_path));
    }
}
