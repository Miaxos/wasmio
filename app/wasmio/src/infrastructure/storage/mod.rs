use std::path::PathBuf;

use chrono::{DateTime, Utc};
use futures::future::join;
use tokio::io::{AsyncRead, AsyncWrite};
use tracing::warn;

/// Implement this trait which define the backend storage used to store data
///
/// The storage is very simple for now
pub trait BackendStorage {
    /// To create a new database
    async fn new_database(&self, name: &str) -> anyhow::Result<DatabaseInfo>;

    /// To get database metadata, if None, database doesn't exist
    async fn database_metadata(&self, name: &str) -> anyhow::Result<Option<DatabaseInfo>>;

    /// List elements from the database,
    async fn list_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        start_after: Option<&str>,
        name_elt: &str,
    ) -> anyhow::Result<ElementInfo>;

    /// Get element from the database,
    async fn get_element_in_database<T: AsyncWrite + Send + Unpin, S: AsRef<str>>(
        &self,
        db: &str,
        key: &str,
        writer: &mut T,
    ) -> anyhow::Result<Option<ElementInfo>>;

    /// Get element metadata from the database,
    async fn get_element_metadata_in_database<T: AsyncWrite + Send + Unpin, S: AsRef<str>>(
        &self,
        db: &str,
        key: &str,
    ) -> anyhow::Result<Option<ElementInfo>>;

    /// Put an element inside database
    async fn insert_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        name_elt: &str,
        content: R,
    ) -> anyhow::Result<ElementInfo>;

    /// Put an element inside database
    async fn delete_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        name_elt: &str,
    ) -> anyhow::Result<ElementInfo>;
}

/// List of database info available
#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct DatabaseInfo {
    name: String,
    number_element: u64,
    created_at: DateTime<Utc>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct ElementInfo {
    name: String,
    size: usize,
    created_at: DateTime<Utc>,
}

// -------------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub struct FSStorage {
    base_path: PathBuf,
}

impl FSStorage {
    pub fn new(base: PathBuf) -> Self {
        Self { base_path: base }
    }
}

impl BackendStorage for FSStorage {
    async fn new_database(&self, name: &str) -> anyhow::Result<DatabaseInfo> {
        let db_info = DatabaseInfo {
            name: name.to_string(),
            number_element: 0,
            created_at: Utc::now(),
        };
        let write_metadata = tokio::fs::write(
            self.base_path.join(format!("{name}.meta")),
            serde_json::to_string(&db_info)?,
        );
        let write_dir = tokio::fs::create_dir(self.base_path.join(name));

        let (a, b) = join(write_metadata, write_dir).await;
        a?;
        b?;

        Ok(db_info)
    }

    async fn database_metadata(&self, name: &str) -> anyhow::Result<Option<DatabaseInfo>> {
        let ressource_path = self.base_path.join(format!("{name}.meta"));

        if let Err(err) = tokio::fs::metadata(&ressource_path).await {
            warn!("{err:?}");
            return Ok(None);
        }

        let content = tokio::fs::read_to_string(ressource_path).await?;

        let data_info = serde_json::from_str(&content)?;
        Ok(Some(data_info))
    }

    async fn insert_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        name_elt: &str,
        content: R,
    ) -> anyhow::Result<ElementInfo> {
        unimplemented!()
    }

    async fn get_element_in_database<T: AsyncWrite + Send + Unpin, S: AsRef<str>>(
        &self,
        db: &str,
        key: &str,
        writer: &mut T,
    ) -> anyhow::Result<Option<ElementInfo>> {
        unimplemented!()
    }

    /// Get element metadata from the database,
    async fn get_element_metadata_in_database<T: AsyncWrite + Send + Unpin, S: AsRef<str>>(
        &self,
        db: &str,
        key: &str,
    ) -> anyhow::Result<Option<ElementInfo>> {
        unimplemented!()
    }

    async fn list_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        start_after: Option<&str>,
        name_elt: &str,
    ) -> anyhow::Result<ElementInfo> {
        unimplemented!()
    }

    async fn delete_element_in_database<R: AsyncRead + Unpin>(
        &self,
        db: &str,
        name_elt: &str,
    ) -> anyhow::Result<ElementInfo> {
        unimplemented!()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[tokio::test]
    async fn simple_db_with_fs() {
        let temp = tempdir().unwrap();
        let fs = FSStorage::new(temp.path().to_path_buf());

        let result = fs.new_database("test").await;
        assert!(result.is_ok());

        let check_metadata_info = fs.database_metadata("test").await;
        assert!(check_metadata_info.is_ok());
        let check_metadata_info = check_metadata_info.unwrap();
        assert!(check_metadata_info.is_some());
    }
}
