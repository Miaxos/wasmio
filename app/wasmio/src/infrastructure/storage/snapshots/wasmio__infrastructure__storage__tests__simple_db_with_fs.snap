---
source: app/wasmio/src/infrastructure/storage/mod.rs
assertion_line: 166
expression: result.unwrap()
---
DatabaseInfo {
    name: "test",
    number_element: 0,
    created_at: 2024-02-08T16:08:56.582873Z,
}
